File: src/TrackForge/Core/MatchCandidate.cs
namespace TrackForge.Core;

/// <summary>One possible metadata match returned by an online source.</summary>
public sealed class MatchCandidate
{
    public string Source { get; set; } = "";
    public string Title { get; set; } = "";
    public string Artist { get; set; } = "";
    public string Album { get; set; } = "";
    public string AlbumArtist { get; set; } = "";
    public string Year { get; set; } = "";
    public string Genre { get; set; } = "";
    public string TrackNumber { get; set; } = "";
    public string TrackCount { get; set; } = "";
    public string DiscNumber { get; set; } = "";
    public string Isrc { get; set; } = "";
    public string Publisher { get; set; } = "";
    public string Bpm { get; set; } = "";
    public int DurationSeconds { get; set; }
    public string ArtUrl { get; set; } = "";
    public string ArtThumbUrl { get; set; } = "";
    public string? AlbumId { get; set; }
    public double Score { get; set; }

    public string Display =>
        $"{Artist} - {Title}" +
        (string.IsNullOrWhiteSpace(Album) ? "" : $"  [{Album}") +
        (string.IsNullOrWhiteSpace(Year) ? "" : $" {Year}") +
        (string.IsNullOrWhiteSpace(Album) ? "" : "]") +
        $"  ({Source} {Score:0})";

    /// <summary>Copy the populated fields onto a track.</summary>
    public void ApplyTo(Track t, bool overwrite, bool titleCase, IEnumerable<string>? only = null)
    {
        var allow = only is null ? null : new HashSet<string>(only, StringComparer.OrdinalIgnoreCase);
        bool Want(string field, string current, string incoming) =>
            !string.IsNullOrWhiteSpace(incoming)
            && (allow is null || allow.Contains(field))
            && (overwrite || string.IsNullOrWhiteSpace(current));

        string Case(string s) => titleCase ? NameFormatter.TitleCase(s) : s;

        if (Want("title", t.Title, Title)) t.Title = Case(Title);
        if (Want("artist", t.Artist, Artist)) t.Artist = Case(Artist);
        if (Want("albumartist", t.AlbumArtist, AlbumArtist)) t.AlbumArtist = Case(AlbumArtist);
        if (Want("album", t.Album, Album)) t.Album = Case(Album);
        if (Want("year", t.Year, Year)) t.Year = Year;
        if (Want("genre", t.Genre, Genre)) t.Genre = Genre;
        if (Want("track", t.TrackNumber, TrackNumber)) t.TrackNumber = TrackNumber;
        if (Want("trackcount", t.TrackCount, TrackCount)) t.TrackCount = TrackCount;
        if (Want("disc", t.DiscNumber, DiscNumber)) t.DiscNumber = DiscNumber;
        if (Want("isrc", t.Isrc, Isrc)) t.Isrc = Isrc;
        if (Want("publisher", t.Publisher, Publisher)) t.Publisher = Publisher;
        if (Want("bpm", t.Bpm, Bpm)) t.Bpm = Bpm;
    }
}

File: macOS/Sources/MatchCandidate.swift
import Foundation

/// One possible metadata match returned by an online source.
final class MatchCandidate: Identifiable, @unchecked Sendable {
    let id = UUID()

    var source = ""
    var title = ""
    var artist = ""
    var album = ""
    var albumArtist = ""
    var year = ""
    var genre = ""
    var trackNumber = ""
    var trackCount = ""
    var discNumber = ""
    var isrc = ""
    var publisher = ""
    var bpm = ""
    var durationSeconds = 0
    var artURL = ""
    var artThumbURL = ""
    var albumID: String?
    var score: Double = 0

    /// Extra sources that donated fields when this is a merged result.
    var mergedFrom: [String] = []

    init() {}

    var sourceLabel: String {
        mergedFrom.isEmpty ? source : "\(source) + \(mergedFrom.joined(separator: " + "))"
    }

    var display: String {
        var s = "\(artist) - \(title)"
        if !album.isBlank { s += "  [\(album)" }
        if !year.isBlank { s += " \(year)" }
        if !album.isBlank { s += "]" }
        s += String(format: "  (%@ %.0f)", source, score)
        return s
    }

    /// Copy the populated fields onto a track.
    func apply(to t: Track, overwrite: Bool, titleCase: Bool, only: [String]? = nil) {
        let allow = only.map { Set($0.map { $0.lowercased() }) }

        func want(_ field: String, _ current: String, _ incoming: String) -> Bool {
            guard !incoming.isBlank else { return false }
            if let allow, !allow.contains(field) { return false }
            return overwrite || current.isBlank
        }
        func cased(_ s: String) -> String { titleCase ? NameFormatter.titleCase(s) : s }

        if want("title", t.title, title) { t.title = cased(title) }
        if want("artist", t.artist, artist) { t.artist = cased(artist) }
        if want("albumartist", t.albumArtist, albumArtist) { t.albumArtist = cased(albumArtist) }
        if want("album", t.album, album) { t.album = cased(album) }
        if want("year", t.year, year) { t.year = year }
        if want("genre", t.genre, genre) { t.genre = genre }
        if want("track", t.trackNumber, trackNumber) { t.trackNumber = trackNumber }
        if want("trackcount", t.trackCount, trackCount) { t.trackCount = trackCount }
        if want("disc", t.discNumber, discNumber) { t.discNumber = discNumber }
        if want("isrc", t.isrc, isrc) { t.isrc = isrc }
        if want("publisher", t.publisher, publisher) { t.publisher = publisher }
        if want("bpm", t.bpm, bpm) { t.bpm = bpm }
    }
}
